### DISCUSSION TITLE
datasets/detect/coco/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9538] [Docs]
# datasets/detect/coco/

Learn how COCO, a leading dataset for object detection and segmentation, integrates with Ultralytics. Discover ways to use it for training YOLO models.

https://docs.ultralytics.com/datasets/detect/coco/

**Comment by FRAOTIAC** [comment]
Thanks for the detailed instructions. Could you please clarify the exact directory structure for the train, val, and annotations folders? Specifically, where should these be placed within the dataset root directory?

**Reply by pderrenger** [reply]
Hello!

For setting up the COCO dataset with the correct directory structure, here's a concise guide:

1. **Dataset Root Directory**: This is your main folder where the dataset will reside, typically named `coco` or similar.
2. **Train, Val, and Test Directories**: Inside the root directory, you should have separate subdirectories for `train2017`, `val2017`, and `test2017`. Each of these will contain the respective images.
3. **Annotations Folder**: This should also be directly inside the root directory. It will contain the annotation files like `instances_train2017.json`, `instances_val2017.json`, etc.

Here’s how it should look:
```
coco/
├── annotations/              # Annotation files
│   ├── instances_train2017.json
│   ├── instances_val2017.json
│   └── ...
├── train2017/                # Training images
├── val2017/                  # Validation images
└── test2017/                 # Test images (optional)
```

Make sure the paths in your `coco.yaml` file correctly point to these directories. If you have any more questions or need further assistance, feel free to ask!

**Reply by casperking666** [reply]
where do you put the coco.yaml file though?

**Reply by pderrenger** [reply]
Hi there!

Great question! The `coco.yaml` file should be placed in a directory where your training script can easily access it. Typically, you would place it in the root directory of your project or in a dedicated `configs` or `datasets` folder within your project structure.

Here’s an example of how your project directory might look:

```
my_yolo_project/
├── coco.yaml
├── datasets/
│   ├── coco/
│   │   ├── annotations/
│   │   │   ├── instances_train2017.json
│   │   │   ├── instances_val2017.json
│   │   │   └── ...
│   │   ├── train2017/
│   │   ├── val2017/
│   │   └── test2017/
├── models/
│   ├── yolov8n.pt
│   └── ...
├── scripts/
│   ├── train.py
│   └── ...
└── ...
```

In your training script, you would reference the `coco.yaml` file like this:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Train the model
results = model.train(data="coco.yaml", epochs=100, imgsz=640)
```

Or via the command line:

```bash
yolo detect train data=coco.yaml model=yolov8n.pt epochs=100 imgsz=640
```

This setup ensures that the `coco.yaml` file is easily accessible and correctly referenced in your training scripts.

If you have any more questions or need further assistance, feel free to ask! 😊

**Comment by aasharma90** [comment]
Hello, 

What is the pre-trained model `yolov8n.pt` trained on? Since if it's already COCO dataset, I wonder why the new training is required.

**Reply by glenn-jocher** [reply]
Hello,

Great question! The pre-trained model `yolov8n.pt` is indeed trained on the COCO dataset. This means it has already learned to detect the 80 object categories included in COCO, such as people, cars, animals, and more.

However, there are several reasons why you might want to retrain or fine-tune this model:

1. **Custom Dataset**: If you have a custom dataset with different object categories or specific variations of the COCO categories, retraining the model will help it adapt to your specific data.

2. **Improved Performance**: Fine-tuning the model on your dataset can improve its performance, especially if your dataset has different characteristics (e.g., different lighting conditions, object sizes, or backgrounds) compared to the original COCO dataset.

3. **Additional Classes**: If you need to detect additional classes that are not included in the COCO dataset, you will need to retrain the model with your extended dataset.

4. **Domain Adaptation**: If your application involves a specific domain (e.g., medical imaging, industrial inspection), retraining the model on domain-specific data can significantly enhance its accuracy and reliability.

Here’s a quick example of how you can fine-tune the `yolov8n.pt` model on your custom dataset using Python:

```python
from ultralytics import YOLO

# Load the pre-trained model
model = YOLO("yolov8n.pt")

# Train the model on your custom dataset
results = model.train(data="path/to/your/custom_dataset.yaml", epochs=100, imgsz=640)
```

And using the CLI:

```bash
yolo detect train data=path/to/your/custom_dataset.yaml model=yolov8n.pt epochs=100 imgsz=640
```

For more details on the COCO dataset and how to use it, you can refer to our [COCO Dataset Documentation](https://docs.ultralytics.com/datasets/detect/coco/).

If you have any more questions or need further assistance, feel free to ask! 😊

**Reply by aasharma90** [reply]
Hi @glenn-jocher 

Thanks for your detailed explanation and reply. 

I understand the fine-tuning requirement for a custom dataset, but my question was about the fine-tuning part mentioned for the COCO dataset itself in the `Train Example` segment above. 

Never mind, I think I have my answers. Thanks so much again :)

**Reply by glenn-jocher** [reply]
Hi @aasharma90,

Thank you for your follow-up! I appreciate your understanding regarding the fine-tuning process.

To clarify, the `Train Example` segment for the COCO dataset is primarily included to demonstrate how to train a YOLO model from scratch or fine-tune it, even if it has already been pre-trained on COCO. This can be useful for several reasons:

1. **Hyperparameter Tuning**: Fine-tuning allows you to adjust hyperparameters specific to your use case, potentially improving model performance.
2. **Extended Training**: Additional training can help the model refine its weights and improve accuracy, especially if you have more computational resources or time.
3. **Custom Augmentations**: You might want to apply custom data augmentations that are more suited to your specific application or environment.

If you have any more questions or need further assistance, feel free to ask! 😊

**Reply by JackEdoubleS** [reply]
I have the same question. If retraining is for domain adaptation and learning new categories, the knowledge from learned model from COCO dataset will be catastrophic forgetting. This means it is the same between retraining with pretrained network and training from scratch.

**Reply by glenn-jocher** [reply]
Thank you for your insights. While catastrophic forgetting is a valid concern, fine-tuning with pre-trained weights often accelerates convergence and improves performance, especially when the new dataset shares similarities with COCO. Training from scratch may be suitable for entirely different domains, but pre-trained models retain valuable feature representations that are not fully overwritten during fine-tuning. Let me know if you'd like more resources or examples to explore this further.

**Reply by JackEdoubleS** [reply]
So right now I need to run YOLO on real world application, but I only have coco dataset. Is it good if I train on coco and test on my application?

**Reply by glenn-jocher** [reply]
Running YOLO on real-world applications after training on the COCO dataset can work well if your application involves detecting objects similar to COCO's 80 categories. However, for domain-specific or unseen objects, fine-tuning on a custom dataset is recommended to improve performance. Let me know if you'd like guidance on fine-tuning.

**Reply by JackEdoubleS** [reply]
Yes. Could you please give me the fine-tuning guidance?

**Reply by glenn-jocher** [reply]
To fine-tune the model for your application, load the pretrained model (e.g., `yolo11n.pt`) and train it on your dataset using the `train` mode. Ensure your dataset is formatted correctly (e.g., COCO format or YAML-configured). For example:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")  # Load the pretrained model
results = model.train(data="your_dataset.yaml", epochs=50, imgsz=640)
```

This process adapts the pretrained weights to your specific use case. For more, refer to the [Ultralytics training documentation](https://docs.ultralytics.com/modes/train/). Let me know if you need further assistance!

**Comment by konstantinos-p** [comment]
It is very unclear from your docs how to finetune your models. In particular, I would expect an example in the quickstart for the case of finetuning on a custom dataset. It is still unclear to me what should be the format of the labels of the finetuning dataset. I've seen some answers in the comments about the labels being "like COCO" but what does that even mean?

**Reply by konstantinos-p** [reply]
Ok, I've found the relevant page in the docs. I still think that it takes way to long to reach this page. Please consider adding some link in the quickstart.

**Reply by pderrenger** [reply]
@konstantinos-p thank you for your feedback! We appreciate your suggestion to make the documentation more accessible, especially regarding finetuning on custom datasets. 

To clarify, finetuning a YOLO model on a custom dataset involves a few key steps:

1. **Dataset Preparation**: Your custom dataset should be formatted similarly to the COCO dataset. This means having images and corresponding annotation files. The annotations should be in JSON format, containing bounding boxes, segmentation masks, or keypoints, depending on your task.

2. **Dataset YAML Configuration**: Create a YAML file that specifies the paths to your training, validation, and test datasets, as well as the class names. Here's an example of what this might look like:
    ```yaml
    path: ../datasets/my_custom_dataset
    train: images/train
    val: images/val
    test: images/test

    # Classes
    names:
      0: class1
      1: class2
      ...
    ```

3. **Training the Model**: Use the `train` command to start training your model on the custom dataset. Here’s an example in Python and CLI:
    ```python
    from ultralytics import YOLO

    # Load a pretrained model
    model = YOLO("yolov8n.pt")

    # Train the model on your custom dataset
    results = model.train(data="path/to/your_custom_dataset.yaml", epochs=100, imgsz=640)
    ```

    ```bash
    yolo detect train data=path/to/your_custom_dataset.yaml model=yolov8n.pt epochs=100 imgsz=640
    ```

For more detailed information on the COCO dataset format and how to structure your custom dataset, please refer to our [COCO Dataset Documentation](https://docs.ultralytics.com/datasets/detect/coco/).

We will consider your suggestion to add more direct links and examples in the quickstart guide to improve accessibility. Thank you for helping us enhance our documentation! If you have any more questions or need further assistance, feel free to ask. 😊

**Comment by TanWeeKang** [comment]
Hi,  i used cvat to annotate my images and saved it as coco format. Then i need to convert it to yolov8 format for training. My question is what structure is the dataset format after converting to yolov8 format? From what i see, there will be images, annotations, labels, and yaml file. I dont understand what is that labels file, is it necessary to have in order to train the dataset using yolov8? Can you explain to me what is the labels file?

**Reply by pderrenger** [reply]
@TanWeeKang hi there!

Great to hear that you're working with CVAT and YOLOv8! When converting your COCO annotations to the YOLOv8 format, the dataset structure will indeed include images, annotations, labels, and a YAML file. Here's a brief overview of each component:

1. **Images**: This folder contains all the images used for training, validation, and testing.
2. **Annotations**: This folder typically holds the original COCO annotations in JSON format.
3. **Labels**: This is a crucial component for YOLO training. The labels folder contains text files where each file corresponds to an image and includes the bounding box coordinates and class labels in YOLO format.
4. **YAML File**: This file (e.g., `coco.yaml`) defines the dataset configuration, including paths to the images and labels, and the class names.

### Understanding the Labels File
The labels file is essential for training YOLO models. Each text file in the labels directory corresponds to an image and contains lines of annotations in the following format:
```
class_id x_center y_center width height
```
- `class_id`: The class index of the object.
- `x_center`, `y_center`: The normalized center coordinates of the bounding box (values between 0 and 1).
- `width`, `height`: The normalized width and height of the bounding box (values between 0 and 1).

### Example
For an image `image1.jpg`, the corresponding label file `image1.txt` might look like this:
```
0 0.5 0.5 0.2 0.4
1 0.3 0.7 0.1 0.2
```
This indicates two objects: one of class 0 and another of class 1, with their respective bounding box coordinates.

### Conversion Tools
To convert COCO annotations to YOLO format, you can use various tools and scripts available online. One such tool is the `coco2yolo` script, which can help automate this process.

### Training Example
Once your dataset is structured correctly, you can train your YOLOv8 model using the following command:
```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Train the model
results = model.train(data="coco.yaml", epochs=100, imgsz=640)
```

For more detailed information, you can refer to the [Ultralytics COCO Dataset Documentation](https://docs.ultralytics.com/datasets/detect/coco/).

If you have any further questions or need additional assistance, feel free to ask. Happy training! 🚀

**Comment by dordanin** [comment]
how can i train the model only on 5 classes from COCO data set ?

**Reply by glenn-jocher** [reply]
@dordanin to train the model on only 5 classes from the COCO dataset, modify the `coco.yaml` file to include only the desired classes in the `names` section and update the `train` and `val` paths accordingly. Then, use this custom YAML file for training. For more details, visit [COCO Dataset Documentation](https://docs.ultralytics.com/datasets/detect/coco/).

**Reply by Hoozt** [reply]
This does not work anymore, the training script labels the images as corrupt and are ignored during training.

**Comment by mod777** [comment]
Hello,

I want to add Goat and give class ID 80 into COOC.yaml data. set Would you mind to tell me in which file (s) i have to changes and  how can i add that goat into coco.yaml data set with label and annotation etc etc

I have tried few times but totally unluck detail or step by step would be great full for me to understand. 
thanks

**Reply by glenn-jocher** [reply]
To add a new class like "Goat" to the COCO dataset, you'll need to modify the `coco.yaml` file by adding the class under the `names` section with the desired ID. Additionally, ensure your dataset includes the corresponding labels and annotations for this new class. For detailed guidance on dataset preparation, please refer to the Ultralytics documentation.

**Reply by mod777** [reply]
I have added to goat and deer into coco.yaml file also have added them in annotated files both instances_train2017.json and instances_val2017.json

	{
            "supercategory": "animal",
            "id": 91,
            "name": "goat"
        },
        {
            "supercategory": "animal",
            "id": 92,
            "name": "deer"
        },
 this is a python code to check 
Classes in Training JSON:
ID: 91, Name: goat, Supercategory: animal
ID: 92, Name: deer, Supercategory: animal

Classes in Validation JSON:
ID: 91, Name: goat, Supercategory: animal
ID: 92, Name: deer, Supercategory: animal

Classes in COCO YAML:
ID: 91, Name: goat
ID: 92, Name: deer

Additional Information for Goat (ID 91):
Annotation for ID 91: BBox: [100, 200, 50, 50], Area: 2500
Annotation for ID 91: BBox: [30.5, 65.3, 30.0, 30.0], Area: 2000.3

Additional Information for Deer (ID 92):
Annotation for ID 92: BBox: [150, 250, 60, 60], Area: 3600
Annotation for ID 92: BBox: [45.0, 30.5, 35.5, 40.0], Area: 2500.5

after all these i cant see goat when i test and check my model. as 91: goat

can u please help me? if u want i give give files so u can have a llok is possible

**Reply by mod777** [reply]
i want to use coco.yaml file class  as well as only to add goat in there nothing else

**Reply by glenn-jocher** [reply]
To add "Goat" to the COCO dataset, ensure you've updated both the `coco.yaml` file and the JSON annotation files correctly. If the class isn't appearing during testing, double-check that the model is trained with the updated dataset and verify the annotations are correctly formatted. If issues persist, consider reviewing the dataset preparation steps or consulting the Ultralytics documentation for further guidance.

**Reply by mod777** [reply]
Hello glenn, 

thanks for your reply. 

I have change the annotation files instances_train2017.json and instances_val2017.json to assign annotation filed for goat 80 to following fileds

Annotation ID: 
Image ID: 
Segmentation: 
Area: 
Iscrowd: 
Category ID: 

also add 80: goat at coco.yaml file

now what i should do next to recognize its goat by coco.yaml or yolov5?
thanks

**Reply by mod777** [reply]
I have change the annotation files instances_train2017.json and instances_val2017.json to assign annotation filed for goat 80 to following fileds

Annotation ID:
Image ID:
Segmentation:
Area:
Iscrowd:
Category ID: 

this assigned is done using pythong code.  is it ok

**Reply by mod777** [reply]
its really too hard to incorporate goat into existing coco.yaml file. i have spend long time to do but no luck. Is there any way to add goat only one object into coco.yaml. coz i have to use defauly cooc.yaml file no custom file. but its really hard. Is there a way out?

**Reply by mod777** [reply]
from ultralytics import YOLO

# Load a pretrained model
model = YOLO("yolov5x.pt")

# Train the model with your custom data
results = model.train(data="coco.yaml", epochs=100, imgsz=640)

this code is not works for me. i m tired. please help

**Reply by glenn-jocher** [reply]
I'm sorry to hear you're having trouble. Please ensure your dataset paths and annotations are correctly set up in the `coco.yaml` file and that your custom data is formatted correctly. If issues persist, reviewing the Ultralytics documentation or community forums may provide additional insights.

**Reply by mod777** [reply]
I have seen few google documentation as well as Ultralytics documentation but everywhere is talk about how to set or train your owned object detection. I can't find anywhere how to include my object into coco.yaml data set. can u give me link for that? thx

**Reply by glenn-jocher** [reply]
To add a new class like "Goat" to the COCO dataset, ensure you've updated the `coco.yaml` file and the JSON annotation files correctly. Unfortunately, there isn't a specific guide for adding classes to COCO, but you can refer to general dataset preparation tutorials. If you encounter issues, consider checking community forums for additional support.

**Reply by mod777** [reply]
If you encounter issues, consider checking community forums for additional support.

can u recommend few?

**Reply by glenn-jocher** [reply]
For community support, consider exploring forums like Stack Overflow, Reddit's r/computervision, or the Ultralytics YOLO GitHub Discussions page. These platforms can provide valuable insights and assistance from other users and developers.

**Comment by xd775** [comment]
In your code example for how the .yaml should be set up: 
```
path: ../datasets/my_custom_dataset
train: images/train
val: images/val
test: images/test

# Classes
names:
  0: class1
  1: class2
  ... 
```

How would I specify which of my annotation .jsons to use for the train and val datasets? 
Also, this thread is confusing me about whether YOLO can read in COCO datasets or if converting to another format is required.

**Reply by glenn-jocher** [reply]
@xd775 for COCO-formatted datasets, you don't need to specify JSON paths in the YAML - the annotations are automatically associated with images based on standard COCO directory structure. YOLO11 directly supports COCO datasets without conversion when using the provided coco.yaml configuration. For custom datasets, we recommend converting to YOLO format using our conversion tools as shown in the COCO dataset documentation.
